woocommerce/woocommerce-ios
# Aztec editor for Product description: part 2.1 - format toolbar with simple actions

Part 2.1 for #1421 

## Changes

Note: I plan on adding unit tests in a separate subtask in #1421

- Created a protocol `AztecFormatBarCommand` to represent a command that can handle a formatting action
- Created `AztecFormatBarCommandCoordinator` that coordinates a given collection of commands, internally stores a map from the formatting identifier to the command that handles the formatting. There are 11 formatting commands in total, and 8 simple formatting commands implemented in this PR. The remaining 3 commands (ordered/unordered list, heading, link) are to be implemented in a followup PR as Part 2.2 of #1421 (you can see the code in this commit: https://github.com/woocommerce/woocommerce-ios/commit/16e642f2611d3fd3f4f8b9cc18734d256817be51).
  - `AztecBoldFormatBarCommand`: bold text formatting
  - `AztecItalicFormatBarCommand`: italic text formatting
  - `AztecUnderlineFormatBarCommand`: underline text formatting
  - `AztecStrikethroughFormatBarCommand`: strikethrough text formatting
  - `AztecBlockquoteFormatBarCommand`: blockquote text formatting
  - `AztecHorizontalRulerFormatBarCommand`: inserting a horizontal ruler
  - `AztecInsertMoreFormatBarCommand`: inserting "more" element
  - `AztecSourceCodeFormatBarCommand`: switching between HTML and Rich Text
- Created `AztecFormatBarFactory` to generate:
  - Format bar (`Aztec.FormatBar`) that is displayed above the keyboard with formatting options
  - Format bar action handler (`AztecFormatBarCommandCoordinator`)
- Added `Aztec.FormatBar` extension `update(editorView: EditorView)` to update the format bar to highlight the formatting options applied to the text being edited
- Created a protocol `FormatBarItemViewProperties` as properties to format a format bar item, and implemented this in the supported `FormattingIdentifier` cases
- Provided mapping from Aztec `TextList` (unordered/ordered list) and `Header` cases to a formatting identifier

## Testing
- Launch the app
- Go to Products tab
- Tap on a Product
- Edit the text --> a toolbar should appear above the keyboard, with 11 formatting options
- Try playing with the "bold" formatting option "B" --> any text can be bolded/unbolded, and when the cursor is at bolded text, "B" should be selected in the format bar
- Try playing with the "italic" formatting option "I" --> any text can be italic'ed/un-italic'ed, and when the cursor is at a text italic'ed, "I" should be selected in the format bar
- Try playing with the "underline" formatting option "U" --> any text can be underlined or not, and when the cursor is at an underlined text, "U" should be selected in the format bar
- Try playing with the "blockquote" formatting option "**“**" --> any text can be block quoted or not, and when the cursor is at a quoted text, "**“**" should be selected in the format bar
- Try playing with the "strikethrough" formatting option "-S-" --> any text can be stricken through or not, and when the cursor is at a stricken through text, "-S-" should be selected in the format bar
- Try inserting a "more" separator --> it should be able to be inserted anywhere  <img width="36" alt="Screen Shot 2019-11-08 at 11 38 21 AM" src="https://user-images.githubusercontent.com/1945542/68447831-4f3da000-021c-11ea-936b-fb7d0a60d9b7.png">
- Try inserting a horizontal separator with the formatting option "—" --> it should be able to be inserted anywhere
- Try switching between HTML and Rich Text mode with the formatting option "</>" --> all the other formatting options should be disabled in HTML mode
- Tap "Done" on the Product description edit screen --> check on web that the description is saved with the above formatting

## Example screenshots

bold/blockquoted/underlined text | source code
-- | --
![Simulator Screen Shot - iPhone SE - 2019-11-08 at 11 32 30](https://user-images.githubusercontent.com/1945542/68447645-9d9e6f00-021b-11ea-9b39-3b14e605be07.png) | ![Simulator Screen Shot - iPhone SE - 2019-11-08 at 11 33 11](https://user-images.githubusercontent.com/1945542/68447646-9d9e6f00-021b-11ea-84ab-e1c4dad3fd3c.png)


Update release notes:
